shell/app-shell/elements/shell-ui.js

```diff
@@ -210,6 +211,75 @@ class ShellUi extends Xen.Debug(Xen.Base, log) {
   _onShare(e, share) {
     this._fire('share', share);
   }
+  _onMicStart() {
+    this._setState({micHidden: false});
+  }
+  _onMicEnd(e) {
+    this._setState({micHidden: true});
+    const finalTranscript = e.detail.toLowerCase().replace(/[^A-Za-z0-9 ]/g, '');
+    log(`voice trigger: [${finalTranscript}]`);
+    if (!finalTranscript) {
+      return;
+    }
+    // see if `node` is a better match for `search` than `match`
+    const findMatch = (node, trigger, search, match) => {
+      log(`node "${node.localName}" has trigger "${trigger}"`);
+      if (trigger.includes(search)) {
+        const diff = trigger.length - search.length;
+        if (!match || match.diff > diff) {
+          log(`matched "${trigger}"`);
+          match = {diff, node};
+        }
+      }
+      return match;
+    };
+    // find all nodes with `trigger` attribute
+    let nodes = document.querySelectorAll('[trigger]');
+    nodes = Array.from(nodes).concat(Array.from(this.host.querySelectorAll('[trigger]')));
+    // find the node that contains `finalTranscript` with the least number of non-matching characters
+    let match;
+    for (const node of nodes) {
+      const trigger = node.getAttribute('trigger').toLowerCase().replace(/[^A-Za-z0-9 ]/g, '');
+      match = findMatch(node, trigger, finalTranscript, match);
+    }
+    // if we matched, value is the transcript
+    if (match) {
+      match.value = finalTranscript;
+    }
+    // if not, look for a prefix-match
+    else {
+      for (const node of nodes) {
+        const trigger = node.getAttribute('trigger').toLowerCase().replace(/[^A-Za-z0-9 ]/g, '');
+        if (finalTranscript.startsWith(trigger)) {
+          match = {
+            node: node,
+            value: finalTranscript.slice(trigger.length).trim(),
+          };
+          break;
+        }
+      }
+    }
+    // if there is a matching node, install value and click it
+    if (match) {
+      match.node.value = match.value;
+      match.node.click();
+      return;
+    }
+    // find a suggestion matching finalTranscript?
+    const suggestions = document.querySelectorAll('suggestion-element');
+    match = null;
+    for (const suggestion of suggestions) {
+      const trigger = suggestion.textContent.toLowerCase();
+      match = findMatch(suggestion, trigger, finalTranscript, match);
+    }
+    // if there is a matching suggestion, click it
+    if (match) {
+      match.node.click();
```

Comment:
Similar, logging something to note we did this case, and the fallback below if that hits, could be worthwhile.

Reply:
I'm going to remove some logging, add some logging, try to optimize ROI.